File: cs/Problems/Day5.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/5
public sealed class Day5 : IProblem<int>
{
    public int Solve(string input) => SumMiddlePageNumbersOptimized(input);

    private readonly Dictionary<int, List<int>> orderRules = []; // ;)

    public int SumMiddlePageNumbersOptimized(ReadOnlySpan<char> input)
    {
        Span<Range> fileSections = stackalloc Range[2];
        input.Split(fileSections, InputReader.NewLine + InputReader.NewLine);

        var orderingSection = input[fileSections[0]];
        var updatesSection = input[fileSections[1]];
        orderRules.Clear();

        var orderingEnumerator = orderingSection.Split(InputReader.NewLine);
        while (orderingEnumerator.MoveNext())
        {
            var orderingRule = orderingSection[orderingEnumerator.Current];
            int separatorInx = orderingRule.IndexOf('|');
            int greater = int.Parse(orderingRule[..separatorInx]);
            int lower = int.Parse(orderingRule[(separatorInx + 1)..]);

            if (!orderRules.ContainsKey(greater))
                orderRules.Add(greater, []);

            orderRules[greater].Add(lower);
        }

        int middleDigitSum = 0;

        Span<int> pageUpdate = stackalloc int[30];
        var updatesEnumerator = updatesSection.Split(InputReader.NewLine);
        while (updatesEnumerator.MoveNext())
        {
            var pageUpdateStr = updatesSection[updatesEnumerator.Current];
            int actualLength = InputParser.ParseNumbers(pageUpdate, pageUpdateStr, ',');
            bool isOrdered = IsOrdered(pageUpdate, actualLength, orderRules);

            if (isOrdered)
                middleDigitSum += pageUpdate[actualLength / 2];
        }

        return middleDigitSum;
    }

    private static int SumMiddlePageNumbers(string input)
    {
        var split = input.Split(InputReader.NewLine + InputReader.NewLine);
        var pageOrdering = split[0].Split(InputReader.NewLine);
        var pageUpdates = split[1].Split(InputReader.NewLine);

        var orderRules = new Dictionary<int, List<int>>();

        foreach (var orderingRuleStr in pageOrdering)
        {
            int separatorInx = orderingRuleStr.IndexOf('|');
            int greater = int.Parse(orderingRuleStr[..separatorInx]);
            int lower = int.Parse(orderingRuleStr[(separatorInx + 1)..]);

            if (!orderRules.ContainsKey(greater))
                orderRules.Add(greater, []);

            orderRules[greater].Add(lower);
        }

        int middleDigitSum = 0;

        foreach (var pageUpdateStr in pageUpdates)
        {
            var pageUpdate = pageUpdateStr.Split(',').Select(int.Parse).ToArray();
            bool isOrdered = IsOrdered(pageUpdate, pageUpdate.Length, orderRules);

            if (isOrdered)
                middleDigitSum += pageUpdate[pageUpdate.Length / 2];
        }

        return middleDigitSum;
    }

    private static bool IsOrdered(ReadOnlySpan<int> pageUpdate, int length, Dictionary<int, List<int>> orderRules)
    {
        for (int i = 0; i < length - 1; i++)
        {
            int current = pageUpdate[i];

            if (!orderRules.ContainsKey(current))
                return false;

            for (int j = i + 1; j < length; j++)
            {
                int against = pageUpdate[j];

                if (!orderRules[current].Contains(against))
                    return false;
            }
        }

        return true;
    }
}

File: go/problems/day5.go
package problems

import (
	"slices"
	"strconv"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day5 struct{}

func (d Day5) Solve(input string) int {
	split := strings.Split(input, shared.NewLine+shared.NewLine)
	ordering := strings.Split(split[0], shared.NewLine)
	updates := strings.Split(split[1], shared.NewLine)

	rules := make(map[int][]int)

	for _, o := range ordering {
		split := strings.Split(o, "|")
		greater, _ := strconv.Atoi(split[0])
		lower, _ := strconv.Atoi(split[1])

		if _, exists := rules[greater]; !exists {
			rules[greater] = []int{lower}
		} else {
			rules[greater] = append(rules[greater], lower)
		}
	}

	r := 0

	for _, u := range updates {
		split := strings.Split(u, ",")
		pages := make([]int, len(split))

		for i, sp := range split {
			pages[i], _ = strconv.Atoi(sp)
		}

		if pagesOrdered(pages, rules) {
			r += pages[len(pages)/2]
		}
	}

	return r
}

func pagesOrdered(pages []int, rules map[int][]int) bool {
	for i := range len(pages) - 1 {
		cp := pages[i]
		r, exists := rules[cp]
		if !exists {
			return false
		}

		for j := i + 1; j < len(pages); j++ {
			np := pages[j]
			if !slices.Contains(r, np) {
				return false
			}
		}
	}
	return true
}
